### ISSUE TITLE
[QUESTION]: Save model at each epoch

### ISSUE THREAD

**Issue by muammar** [chemprop/chemprop#317]
**What are you trying to do?**
I am trying to save the model at certain number of epochs. 

**Previous attempts**
I could not find any argument in the documentation to accomplish it. 

What would be a way to do it? Inspecting the structure of the final directory after training contains the final `.pt`. 

Thanks.

**Comment by kevingreenman** [comment]
The model that's saved to the `.pt` file is the one that has the best score on the validation set after the number of epochs specified using `--epochs` (so the saved model may or may not be from the final epoch, depending on whether your model has started to overfit on the training set). The score is calculated with the function specified by `--metric`. Are you trying to save a model after an epoch where it doesn't have the best validation-set performance?

**Comment by muammar** [comment]
> The model that's saved to the `.pt` file is the one that has the best score on the validation set after the number of epochs specified using `--epochs` (so the saved model may or may not be from the final epoch, depending on whether your model has started to overfit on the training set). The score is calculated with the function specified by `--metric`. Are you trying to save a model after an epoch where it doesn't have the best validation-set performance? If so, this functionality isn't currently available in Chemprop.

Correct, I want to save the model after each epoch regardless of it being the best or the worst over the validation-set performance. One could easily achieve this when chemprop migrates to work as a python library. In the interim, It would be interesting to have that option implemented because it would let you, for instance, monitor how the predictions or embeddings are adjusted after each epoch for visualization purposes.

**Comment by muammar** [comment]
@shihchengli thanks, that's helpful. I will play with it, and if it is not in the roadmap, I can try preparing a PR.

**Comment by shihchengli** [comment]
I don't think we plan to support this for general users, since most people are only interested in how to get a good performance model. @kevingreenman, I would like to know your opinion on this.

**Comment by muammar** [comment]
> I don't think we plan to support this for general users, since most people are only interested in how to get a good performance model. @kevingreenman, I would like to know your opinion on this.

I understand, but monitoring the behavior of some aspects of the algorithm at each epoch is interesting for some of the users. The `SummaryWriter()` info that is currently dumped is not very useful. I also know there are plans for Chemprop being a library instead of a command line tool and that would be useful to achieve what I asked in this thread instead of hardcoding. Feel free to close this report.

**Comment by shihchengli** [comment]
@muammar Thanks for this suggestion. We think it doesn't hurt to have this option in Chemprop. Actually, one of the developers has already done this before in a branch. If you are interested in adding this feature and making a PR, that would be great!
